MonkeyWrench.Web.WebService/WebServices.asmx.cs

```diff
@@ -2633,234 +2633,187 @@ public GetBuildInfoResponse GetBuildInfoMultiple (WebServiceLogin login, string
+						DBRevision dep_lane_rev = dependent_lane.FindRevision (db, revision.revision);
+						if (dep_lane_rev == null) {
+							log.ErrorFormat ("Lane {0} ({1}) is dependent on lane {2} ({3}) but it doesn't have revision {4}", lane.lane, lane.id, dependent_lane.lane, dependent_lane.id, revision.revision);
+							continue;
+						}
+						dep_revwork = DBRevisionWork_Extensions.Find (db, dependent_lane, dependent_host, dep_lane_rev);
+						
+						work_files = dep_revwork.GetFiles (db);
+						
+						foreach (DBWorkFile file in work_files) {
+							bool download = true;
+							foreach (string exp in dep.download_files.Split (new char [] { ' ' }, StringSplitOptions.RemoveEmptyEntries)) {
+								if (!System.Text.RegularExpressions.Regex.IsMatch (file.filename, FileUtilities.GlobToRegExp (exp))) {
+									download = false;
+									break;
 								}
-								files_to_download.Add (file);
-								dependent_lanes.Add (dependent_lane);
 							}
+							if (!download)
+								continue;
+							if (files_to_download == null) {
+								files_to_download = new List<DBWorkFile> ();
+								dependent_lanes = new List<DBLane> ();
+							}
+							files_to_download.Add (file);
+							dependent_lanes.Add (dependent_lane);
 						}
 					}
+				}
 
+				List<DBWorkView2> pending_work = revisionwork.GetNextWork (db, lane, masterhost, revision, multiple_work);
 
-					List<DBWorkView2> pending_work = revisionwork.GetNextWork (db, lane, masterhost, revision, multiple_work);
+				if (pending_work == null || pending_work.Count == 0)
+					return response;
 
-					if (pending_work == null || pending_work.Count == 0)
-						continue;
+				var environment_variables = new List<DBEnvironmentVariable>();
+				using (var cmd = db.CreateCommand ()) {
+					var envVarSet = new HashSet<string> ();
 
-					List<DBEnvironmentVariable> environment_variables = null;
-					using (IDbCommand cmd = db.CreateCommand ()) {
-						foreach (int li in db.GetLaneHierarchy (lane.id)) {
-							cmd.CommandText += string.Format (@"
-SELECT * 
-FROM EnvironmentVariable 
-WHERE 
-(host_id = {0} OR host_id = {1} OR host_id IS NULL) AND (lane_id = {2} OR lane_id IS NULL)
-ORDER BY id;
-;", revisionwork.workhost_id, revisionwork.host_id, li);
-						}
-						using (IDataReader reader = cmd.ExecuteReader ()) {
-							var set = new HashSet<string> ();
-							do {
-								while (reader.Read ()) {
-									if (environment_variables == null)
-										environment_variables = new List<DBEnvironmentVariable> ();
-
-									var ev = new DBEnvironmentVariable (reader);
-									if (!set.Contains (ev.name)) {
-										environment_variables.Add (ev);
-										set.Add (ev.name);
-									}
+					cmd.CommandText = @"
+						SELECT * 
+						FROM EnvironmentVariable 
+						WHERE 
+						(host_id = @host_id OR host_id = @workhost_id OR host_id IS NULL) AND (lane_id = @lane_id OR lane_id IS NULL);
+						--ORDER BY id;
```

Comment:
You're removing the order here, which means that if the same environment variable is defined multiple times, it'll be random which one is chosen.